### PR TITLE
New layout: MonadThreeCol

### PR DESCRIPTION
**Pull request by mbenford**
This PR introduces a new layout: `MonadThreeCol`. It's based on [XMonad's ThreeColumns](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Layout-ThreeColumns.html) layout and it's useful for people using ultra wide displays. It behaves similarly to MonadTall with two key differences: the main window can be placed at the center of the screen and the secondary windows are divided into two columns:

image

Windows are distributed following XMonad's ThreeColumns behavior:

image
image

This is a working in progress and shouldn't be merged yet. It's functional - I've been using it myself - but there are some rough edges that need smoothing out. I wanted to share it as it stands to get some early feedback on both the code (Python isn't my strong suit) and the layout design. I tried to use as much as I could from `MonadTall` and only rewrote the parts that were too coupled to the way the tall layout works. 

To-do list
- [x] Port XMonad's ThreeCol layout.
- [x] Port XMonad's ThreeColMid layout. 
- [x] Fix double margins.
- [x] Fix resize of secondary windows.
- [x] Write tests.

### CODE REVIEW

**Comment by m-col** [comment]
Awesome! Looking at the code itself everything looks good. Do you have an example configuration that someone who hasn't used the Monad layouts (me) could use to test out the different functions and features?

**Comment by mbenford** [comment]
There you go (default config file with custom key bindings for xmonad layout). To be honest, there are no new features besides the layout itself; most configuration options and public methods come from `MonadTall`.

```python
from typing import List  # noqa: F401

from libqtile import bar, layout, widget
from libqtile.config import Click, Drag, Group, Key, Match, Screen
from libqtile.lazy import lazy
from libqtile.utils import guess_terminal

mod = "mod4"
terminal = guess_terminal()

keys = [
    Key([mod], "h", lazy.layout.left(), desc="Move focus to left"),
    Key([mod], "l", lazy.layout.right(), desc="Move focus to right"),
    Key([mod], "j", lazy.layout.down(), desc="Move focus down"),
    Key([mod], "k", lazy.layout.up(), desc="Move focus up"),
    Key([mod], "space", lazy.layout.next(), desc="Move window focus to other window"),

    Key([mod, "shift"], "h", lazy.layout.shuffle_left(), desc="Move window to the left"),
    Key([mod, "shift"], "l", lazy.layout.shuffle_right(), desc="Move window to the right"),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up(), desc="Move window up"),

    Key([mod, "control"], "h", lazy.layout.shrink_main(), desc="Grow window to the left"),
    Key([mod, "control"], "l", lazy.layout.grow_main(), desc="Grow window to the right"),
    Key([mod, "control"], "j", lazy.layout.grow(), desc="Grow window down"),
    Key([mod, "control"], "k", lazy.layout.shrink(), desc="Grow window up"),
    Key([mod], "m", lazy.layout.normalize(), desc="Maximize window"),
    Key([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),

    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),

    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key([mod], "w", lazy.window.kill(), desc="Kill focused window"),

    Key([mod, "control"], "r", lazy.restart(), desc="Restart Qtile"),
    Key([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    Key([mod], "r", lazy.spawncmd(), desc="Spawn a command using a prompt widget"),
]

groups = [Group(i) for i in "123456789"]

for i in groups:
    keys.extend([
        Key([mod], i.name, lazy.group[i.name].toscreen(), desc="Switch to group {}".format(i.name)),
        Key([mod, "shift"], i.name, lazy.window.togroup(i.name, switch_group=True), 
            desc="Switch to & move focused window to group {}".format(i.name)),
    ])

layouts = [
    layout.MonadThreeCol(main_centered=True),
    layout.MonadThreeCol(main_centered=False),
]

widget_defaults = dict(
    font='sans',
    fontsize=12,
    padding=3,
)
extension_defaults = widget_defaults.copy()

screens = [
    Screen(
        bottom=bar.Bar(
            [
                widget.CurrentLayout(),
                widget.GroupBox(),
                widget.Prompt(),
                widget.WindowName(),
                widget.Chord(
                    chords_colors={
                        'launch': ("#ff0000", "#ffffff"),
                    },
                    name_transform=lambda name: name.upper(),
                ),
                widget.TextBox("default config", name="default"),
                widget.TextBox("Press &lt;M-r&gt; to spawn", foreground="#d75f5f"),
                widget.Systray(),
                widget.Clock(format='%Y-%m-%d %a %I:%M %p'),
                widget.QuickExit(),
            ],
            24,
        ),
    ),
]

mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(), start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front())
]

dgroups_key_binder = None
dgroups_app_rules = []  # type: List
follow_mouse_focus = True
bring_front_click = False
cursor_warp = False
floating_layout = layout.Floating(float_rules=[
    *layout.Floating.default_float_rules,
    Match(wm_class='confirmreset'),  # gitk
    Match(wm_class='makebranch'),  # gitk
    Match(wm_class='maketag'),  # gitk
    Match(wm_class='ssh-askpass'),  # ssh-askpass
    Match(title='branchdialog'),  # gitk
    Match(title='pinentry'),  # GPG key password entry
])
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True
auto_minimize = True
wmname = "LG3D"
```

**Comment by mbenford** [comment]
I updated the PR and fixed both double margins and resize of secondary windows. I thought I'd share how I did it because my approaches were different than the ones in `MonadTall`.

**Double margins**

`MonadTall` deals with double margins by adjusting both the y position and the height of all secondary clients other than the first one. Since this new layout can have two columns I tried to fix the problem in a more general way: rightmost and bottommost clients will have all four margins set while other clients will have only north and west margins. This prevents margins of adjacent clients from adding up and all distances among clients will honor the `margin` option. Clients are considered rightmost and/or bottommost when they touch the screen_rect edges (at first I thought this wouldn't work due the position of bars, but qtile takes care of that by providing a corrected screen_rect).

**Resize of secondary windows**

I couldn't use `MonadTall`s resize implementation because it (reasonably) assumes a single column for secondary windows, and I would have to re-implement several methods in `MonadThreeCol`. So I created a single method that handles all resizing cases (grow and shrink) by iterating over all secondary clients in the same column and change their heights accordingly. A single pass should be enough for most cases, and even on edge cases two or three passes will eventually resize everything.

If anyone sees any problems with these approaches, please let me know.

**Comment by mbenford** [comment]
I have started working on the tests and I hope I can get them done soon.

**Comment by jeffzi** [comment]
Hi @mbenford, thanks for this layout :tada: . I've been using it for a couple of months without issues. It's now my default layout on my ultrawide 38"!

I'm not familiar with qtile development but can help with testing if needed.

**Comment by m-col** [comment]
Is this PR finished? Just asking because I see you've added a test for it, but the PR is still marked as a draft.

**Comment by mbenford** [comment]
@m-col I've added a single test so far that ensures windows are correctly placed on the screen. I'd like to write a few more to cover other behaviors as well. But I guess the draft status can be changed at this point since the new layout is 100% done.

@jeffzi I'm glad it's been useful to you. :)

**Comment by m-col** [comment]
OK, no problem. It looks like you'll need to rebase as there is a conflict. Let me know if you'd prefer to merge and add more tests in a future PR, or to hold on merging until then. 

Please could you add an entry to the changelog too? Then others can appreciate your work!

File: libqtile/layout/xmonad.py [hunk 1]
```diff
@@ -1005,3 +1006,290 @@ def cmd_swap_main(self):
...
+                start=0,
+                end=clients[0],
+            ),
+            Column(
+                name="right",
+                count=clients[1],
+                start=clients[0],
+                end=clients[0] + clients[1],
+            )
+        ]
```
Comment (m-col) [inline libqtile/layout/xmonad.py:1277]: Could you make this a tuple? It will always return two columns so makes it more clear.

File: libqtile/layout/xmonad.py [hunk 2]
```diff
@@ -1005,3 +1006,290 @@ def cmd_swap_main(self):
...
+        self.group.layout_all()
+
+    def _get_column(self, index):
+        """Get the column containing the provided index"""
+        left, right = self._get_columns()
+        return left if index < left.count else right
+
+    def _get_columns(self):
+        """Get all columns"""
+        Column = namedtuple("Column", "name count start end")
```
Comment (m-col) [inline libqtile/layout/xmonad.py:1261]: Maybe this should be at the top level of the file so that it isn't recreated repeatedly? Or perhaps as a class attribute.

File: libqtile/layout/xmonad.py [hunk 3]
```diff
@@ -1005,3 +1006,290 @@ def cmd_swap_main(self):
...
+                end=clients[0] + clients[1],
+            )
+        ]
+
+    def _split_integer(self, value, parts):
+        """Divide an integer into equal parts and distribute the remainder"""
+        result = [value // parts] * parts
+        for i in range(value % parts):
+            result[i] += 1
+        return result
```
Comment (m-col) [inline libqtile/layout/xmonad.py:1284]:
```suggestion
    @staticmethod
    def _split_integer(value, parts):
        """Divide an integer into equal parts and distribute the remainder"""
        result = [value // parts] * parts
        for i in range(value % parts):
            result[i] += 1
        return 
```

**Comment by mbenford** [comment]
> OK, no problem. It looks like you'll need to rebase as there is a conflict. Let me know if you'd prefer to merge and add more tests in a future PR, or to hold on merging until then.

@m-col I'll add a few more tests later this week so we can merge this PR more confidently.

**Comment by keslerm** [comment]
Just some noise but I am actively looking forward to this PR!

**Comment by mbenford** [comment]
@m-col I've added a few more tests and I think this PR is good to be merged now. There are still some areas that I want to cover but I can do it later, if that's ok with you.

Also, I'm not sure why the Python 3.10 action failed. Let me know if I need to fix anything on my end.

**Comment by elParaguayo** [comment]
The tests failed the formatting check. You need to run `black`on your modified files to fix this.

**Comment by mbenford** [comment]
@elParaguayo done!

**Comment by mbenford** [comment]
@m-col Do you want me to rebase my branch so there's no merge commit from master?

**Comment by m-col** [comment]
I was just trying to rebase it through github but oddly it says I can't. Not sure why. But yes, that would be great, thanks!

**Comment by m-col** [comment]
Amazing, thank you @mbenford!

**Comment by davidbudzynski** [comment]
Not sure if this is the place to post this, but I think we're missing an icon for this layout in the `CurrentLayoutIcon` widget. I'm on a `qtile-git` version on Arch Linux and  [it currently looks like this.](https://i.imgur.com/adVf4aw.png)

**Comment by m-col** [comment]
Good catch. Would one of you like to put one together @mbenford, or @elParaguayo as you made one recently for `Spiral`?

**Comment by elParaguayo** [comment]
Yeah. I can do it.

**Comment by elParaguayo** [comment]
#3422
